build.gradle

```diff
@@ -166,6 +166,86 @@ subprojects.stream().filter { it.parent != rootProject }.each { p ->
 
         sourceSets {
             main.kotlin.srcDirs += "$projectDir/src"
+            test.kotlin.srcDirs += "$projectDir/test"
+
+            testPlain {
+                kotlin.srcDirs += "$projectDir/test"
+                compileClasspath = test.compileClasspath
+            }
+
+            testCommonJS {
+                kotlin.srcDirs += "$projectDir/test"
+                compileClasspath = test.compileClasspath
+            }
+
+            testUmdOnly {
+                kotlin.srcDirs += "$projectDir/testUMD"
+                compileClasspath = test.compileClasspath
+            }
+
+            testPlainOnly {
+                kotlin.srcDirs += "$projectDir/testPlain"
+                compileClasspath = test.compileClasspath
+            }
+
+            testCommonJsOnly {
+                kotlin.srcDirs += "$projectDir/testCommonJS"
+                compileClasspath = test.compileClasspath
+            }
+
+            // Shared
+
+            testSharedUMD {
+                kotlin.srcDirs += "$projectDir/../_testShared"
+                compileClasspath = test.compileClasspath
+            }
+
+            testSharedPlain {
+                kotlin.srcDirs += "$projectDir/../_testShared"
+                compileClasspath = test.compileClasspath
+            }
+
+            testSharedCommonJS {
+                kotlin.srcDirs += "$projectDir/../_testShared"
+                compileClasspath = test.compileClasspath
+            }
+
+            testSharedUmdOnly {
+                kotlin.srcDirs += "$projectDir/../_testSharedUMD"
+                compileClasspath = test.compileClasspath
+            }
+
+            testSharedPlainOnly {
+                kotlin.srcDirs += "$projectDir/../_testSharedPlain"
+                compileClasspath = test.compileClasspath
+            }
+
+            testSharedCommonJsOnly {
+                kotlin.srcDirs += "$projectDir/../_testSharedCommonJS"
+                compileClasspath = test.compileClasspath
+            }
+        }
+
+        tasks.withType(org.jetbrains.kotlin.gradle.tasks.Kotlin2JsCompile).forEach {
+            if (it.name.startsWith("compileTest") && it.name.endsWith("Kotlin2Js")) {
+
+                def prefix = it.name.startsWith("compileTestShared") ? "compileTestShared" : "compileTest";
+                def start = prefix.length()
+
+                def end = it.name.lastIndexOf("OnlyKotlin2Js")
+                if (end < 0) {
+                    end = it.name.lastIndexOf("Kotlin2Js")
+                }
+
+                def moduleKind = it.name.substring(start, end).toLowerCase()
+
+                if (moduleKind.isEmpty()) {
+                    moduleKind = "umd"
+                }
+
+                it.kotlinOptions.moduleKind = moduleKind
+                test.dependsOn(it)
```

Comment:
IMHO a better solution would be specifying a special property in konfig.json, which tells us which module systems the library is supposed to work with, e.g. `moduleSystems: ['plain', 'commonjs', 'umd']`. I thinks that's much easier to explain and configure.

Reply:
I think both solutions have advantages and disadvantages.
I propose to try to live with this solution now. If we will not like it we will switch to another.

Reply:
Ok)